### PR TITLE
Added new fade and transform animations (:class:`~.TransformMatchingShapes`, :class:`~.TransformMatchingTex`, :class:`~.FadeTransform`) from 3b1b/manim

### PR DESCRIPTION
**Pull request by behackl**
## Motivation

The enhanced transform features illustrated in https://3b1b.github.io/manim/getting_started/example_scenes.html#textransformexample are very useful and should be made available for the community maintained version as well. This PR attempts to do so.

## Overview / Explanation for Changes

In its current form, this just brings over the relevant code from 3b1b/manim and adapts it appropriately:

- `CONFIG` is removed,
- some methods are called differently (`align_submobjects` vs. `align_family`, ...)
- `Tex` strings have a new feature where substrings wrapped in `{{}}` are automatically isolated, e.g., `MathTex({{a^2}} + {{b^2}} = {{c^2}})` essentially renders as `MathTex("a^2", "+", "b^2", "=", "c^2")`.

Added new Fade animation: :class:`~FadeOutToPoint`
Added :class:`~FadeTransform` and :class:`~FadeTransformPieces` for transforming mobjects and submobjects with a fade
Added :class:`~TransformMatchingShapes` and :class:`~TransformMatchingTex` for transforming mobjects and tex that have matching parts

None of this is properly documented yet, help in doing so is appreciated.

## Oneline Summary of Changes

```
- New Feature: TransformMatchingShapes, TransformMatchingTex, and isolation of Tex substrings wrapped in {{...}} (:pr:`PR NUMBER HERE`)
```

## Testing Status
We (still) cannot test animations, but here are some examples of what this implementation can do:

```py
class ExEquation(Scene):
    def construct(self):
        eq1 = MathTex("{{a^2}} + {{b^2}} = {{c^2}}")
        eq2 = MathTex("{{a^2}} = {{c^2}} - {{b^2}}")
        self.add(eq1)
        self.wait(0.5)
        self.play(TransformMatchingTex(eq1, eq2))
        self.wait(0.5)
```
renders as

https://user-images.githubusercontent.com/11851593/108063357-5f693f80-705b-11eb-829a-0bfd0b5a836d.mp4

```py
class ExAnagram(Scene):
    def construct(self):
        src = Text("the morse code")
        tar = Text("here come dots")
        self.play(Write(src))
        self.wait(0.5)
        self.play(TransformMatchingShapes(src, tar, path_arc=PI/2))
        self.wait(0.5)
```
renders as

https://user-images.githubusercontent.com/11851593/108063642-d0a8f280-705b-11eb-8e0e-aa29203cc44e.mp4

## Further Comments

I don't really like that this introduces further `Transform` animations, and I would like to steer this in a more user-friendly direction eventually (maybe where `Transform` becomes more intelligent and tries to apply a more specialized transform based on the arguments the user passes) -- but not for this PR.

## Acknowledgements
- [x] I have read the [Contributing Guidelines](https://docs.manim.community/en/latest/contributing.html)

## Reviewer Checklist
- [ ] Newly added functions/classes are either private or have a docstring
- [ ] Newly added functions/classes have [tests](https://github.com/ManimCommunity/manim/wiki/Testing) added and (optional) examples in the docs
- [ ] Newly added documentation builds, looks correctly formatted, and adds no additional build warnings
- [ ] The oneline summary has been included [in the wiki](https://github.com/ManimCommunity/manim/wiki/Changelog-for-next-release)

### CODE REVIEW

**Comment by behackl** [comment]
I was a bit too optimistic with the changes in `break_up_tex_strings`: the documentation build reports

```
File "/home/docs/checkouts/readthedocs.org/user_builds/manimce/envs/1037/lib/python3.8/site-packages/manim/mobject/graph.py", line 381, in <listcomp>
    [(v, MathTex(v, fill_color=label_fill_color)) for v in vertices]
  File "/home/docs/checkouts/readthedocs.org/user_builds/manimce/envs/1037/lib/python3.8/site-packages/manim/mobject/svg/tex_mobject.py", line 381, in __init__
    tex_strings = self.break_up_tex_strings(tex_strings)
  File "/home/docs/checkouts/readthedocs.org/user_builds/manimce/envs/1037/lib/python3.8/site-packages/manim/mobject/svg/tex_mobject.py", line 408, in break_up_tex_strings
    pieces.extend(re.split(pattern, s))
  File "/home/docs/checkouts/readthedocs.org/user_builds/manimce/envs/1037/lib/python3.8/re.py", line 231, in split
    return _compile(pattern, flags).split(string, maxsplit)
TypeError: expected string or bytes-like object
```

**Comment by kilacoda** [comment]
black is failing.

**Comment by behackl** [comment]
I've finally had some time to polish this PR a bit, it is now ready for review. :-)

Some notes:
- I feel that after the addition of `FadeOutToPoint` we have too many `FadeOut...` animations around, and I would like to see that condensed a little. They probably could all be merged into a single `FadeOut` animation whose behavior can be controlled via keyword arguments.
- Strictly speaking, it wouldn't be necessary for this PR to include the implementations of `FadeTransform` and `FadeTransformPieces`, but I think they are of general interest.
- And in a similar vein, the changes to `break_up_tex_strings` are also not really necessary in order to have the matching shape/tex transformations, but I've found it to be a very useful utility.

**Review by GameDungeon — APPROVED** [review]
LGTM, and nice animations :)

**Review by WampyCakes — APPROVED** [review]
I think that everything introduced in `transform_matching_parts.py` should probably just be added to `transform.py`, but I won't block it for that. It seems there will probably be a cleaning of animations sometime soon anyway. 

Otherwise, it would be nice to have examples of the animations that don't have examples. [The way it's done here](https://docs.manim.community/en/v0.4.0/reference/manim.animation.creation.html#module-manim.animation.creation) is smart since one video covers so many animations. This can come later though. I've actually used these animations quite a bit, and would love to see them in v0.5.0. All of them work for me.

**Review by eulertour — APPROVED** [review]

File: manim/animation/transform.py [hunk 1]
```diff
@@ -401,3 +403,135 @@ def interpolate(self, alpha: float) -> None:
...
+                    FadeTransform(starts[0], targets[0], stretch=True),
+                    FadeTransform(starts[1], targets[1], stretch=False, dim_to_match=0),
+                    FadeTransform(starts[2], targets[2], stretch=False, dim_to_match=1)
+                )
+
+                self.play(*[FadeOut(mobj) for mobj in self.mobjects])
+
+    """
+
+    def __init__(self, mobject, target_mobject, stretch=True, dim_to_match=1, **kwargs):
```
Comment (naveen521kk) [inline manim/animation/transform.py:452]: Can you type this code or at least have typings in docstrings? IIRC, all the code in `animation` is already typed.

**Comment by behackl** [comment]
Thanks for the reviews!

> I think that everything introduced in `transform_matching_parts.py` should probably just be added to `transform.py`, but I won't block it for that. It seems there will probably be a cleaning of animations sometime soon anyway.

I agree that it would fit nicely there -- however, this isn't possible without disentangling the inheritance mess in the `fading` and `transform` modules. I've tried for a bit, but couldn't resolve the circular imports that arise when trying to do so without restructuring the modules significantly.

> Otherwise, it would be nice to have examples of the animations that don't have examples. [The way it's done here](https://docs.manim.community/en/v0.4.0/reference/manim.animation.creation.html#module-manim.animation.creation) is smart since one video covers so many animations. This can come later though. I've actually used these animations quite a bit, and would love to see them in v0.5.0. All of them work for me.

+1, but rather after we have cleaned up the current animation mess.

File: manim/animation/transform.py (hunk 1 above)
Comment (behackl) [reply manim/animation/transform.py:452]: Done via 9cdf699cfff99ac166f901cfd0f71a674668b631.

**Pull request merged**